### ISSUE TITLE
delete / close link preview

### ISSUE THREAD

**Issue by prohde** [RocketChat/Rocket.Chat#1929]
Would it be possible to delete / close a link preview?

I think I'm not the only one who does not want a link preview for every link posted. So deleting it after the post was made would be nice.
Even better would be seeing the preview right while typing the message and being able to close it then. So it would not be appended to the message. (like it is known from Facebook)

**Comment by ant31** [comment]
+1
moreover link preview to site that require auth doesn t make sense(preview the login page).

**Comment by nobeans** [comment]
Agreed. I have a bot notify to a room with some URLs after someone pushes commits to a repository. And many link previews annoy me.

I hope it's customizable in Preferences, like `Use Emojis`.

**Comment by Michuelnik** [comment]
+1
A white-/blacklist for links that are (or aren't) to be previewed could be another approach.

**Comment by geekq** [comment]
If you really annoyed by this feature (like me), as a quick hack, you can replace

    return getDescription(this)? or getTitle(this)?

in the `show` method of
https://github.com/RocketChat/Rocket.Chat/blob/69b3c81840b246e74b2041dfd0f212612fd23d22/packages/rocketchat-oembed/client/oembedUrlWidget.coffee#L40

by 

    return false

so the oembedUrlWidget never shows up.

Or you can try making the condition more intelligent like detecting "sign in" strings:

    return (getDescription(this)? or getTitle(this)?) and (not /sign.?in/i.test(getTitle(this)))

or extend, customize for your usual urls.

The real solution would be to make the widget really intelligent, showing helpful digest of the page content. But it will require much more work and adding heuristics is more art than science (I guess this is the way facebook and slack went).

**Comment by timkinnane** [comment]
Related #5495

**Comment by timkinnane** [comment]
Related #1728

**Comment by mikaelmello** [comment]
@engelgabriel I'm interested in contributing to this, what is the intended design? (Sing Li told me to tag you)

From what I've understood:
 * A preview showing up while you write the message (with an X on it to remove the preview) 
 * After the message is sent, an option to hide the preview on the popover menu if it isn't already hidden
 * A global setting to allow/disallow previews

Is that correct?

**Comment by timkinnane** [comment]
I'm not sure about this part: `A preview showing up while you write the message`
It seems like that could get very tricky to get a good UX balance, especially keeping the feature synced with mobile etc. The display of the preview will be hard to position without it looking messy, with thumbnails and snippets of different dimensions etc and it might obstruct other interface components.

Maybe just something very subtle to indicate that a link is detected and a preview will be added when the message renders in the stream.

On the second point, I definitely think the message author (or anyone who can edit) should be able to remove previews. Maybe even adding an admin config for a URL blacklist, to never load previews from certain domains. That might be helpful for instances with lots of specific links, where the previews just get in the way. e.g. `open.rocket.chat` admins could put GitHub on the list and probably cut down a lot of client loading as well as clear up the reading experience. The repo previews are really not required.

**Comment by timkinnane** [comment]
Illustration of the concept...

![link embed preview](https://user-images.githubusercontent.com/1774379/39902916-df345b34-5513-11e8-9fa8-a60c5274be5e.png)

I think a subtle approach is best. Loading previews while you type could be a very janky visual experience and cause all kinds of client side performance issues. I think the feature would also be only wanted by more advanced users, so it's best to let it surface with some lower priority than the main interfaces, to not overwhelm the basics.

**Comment by mikaelmello** [comment]
On messages with 2 or more URLs, are we going to close all previews or have an X on each oembed to close specific ones?

**Comment by timkinnane** [comment]
It would be nice to close each independently, but it might not be feasible - e.g. if there's a single message property to toggle display of previews, it would be easier to work with in templates etc than something like an array of attachment ids with their own unique toggles.

**Comment by benwick** [comment]
> https://github.com/RocketChat/Rocket.Chat/issues/1929#issuecomment-210478035
A blacklist for which domains shouldn't be previewed would be really great.

**Comment by matortheeternal** [comment]
I think it would be best if the user can simply wrap the link in some characters to indicate it should not be embedded.  This allows the user to easily control which links are embedded and which aren't.  This can be done in addition to a "remove link preview" option.

**Comment by xmedeko** [comment]
@matortheeternal If you add a comma just after the link, the preview is not shown, e.g. `https://gitbug.com,` I am just not sure if it's a feature or a bug :-)

**Comment by geekq** [comment]
Adding comma does not work for all urls, only for some.
More reliable workaround seems to delete the leading `http://` like

```
This shows no preview: www.google.com but is shown as a link in RocketChat
```

**Comment by williamjamir** [comment]
Any news on this issue?
The PR that has an implementation for this issue was closed =/
https://github.com/RocketChat/Rocket.Chat/pull/10755

**Comment by jlm451** [comment]
@williamjamir I'm also curious what happened with that PR. My work uses RC and this issue is a little annoying, I'd be down to try contributing a fix.

**Comment by mikaelmello** [comment]
Sorry guys, I just realized I'd always postpone "the day where I'll finish it" and thought it was best to just close.

I'd be happy if it was implemented too!

**Comment by knokio** [comment]
I have a simple PR with this working, with a simple "Remove Link Preview" message action.

**Comment by knokio** [comment]
Demo: https://streamable.com/p0cxfj

Things to improve:
- Better icon
- Only show option when there is really a preview

Do you have any more suggestions?

**Comment by timoaudi** [comment]
Any update on this? As a user, I would love to have the option to manually remove links per message.

**Comment by Julianoe** [comment]
Just wanted to refresh this topic. I have been hosting a server since 2017 with 3 to 10 users regularly interacting and sharing a lot of links and this would really be a life improvement. 

The `<https://example.com>` would be a really good way of allowing us to not trash a channel with dozens of embeds when we don't want to. This could be added in the little toolbar of shortcuts under the message box input to help users know it's possible.

**Comment by Julianoe** [comment]
The proposed hacks are not satisfying solutions : 
- adding a comma by @xmedeko : does not work for a lot of urls
- or removing `https` by @geekq : you can't use this trick with markdown links because markdown renderer needs https or http to know it's a link and output it like one.

**Comment by knokio** [comment]
Can you check my PR and demo and see if this a desired behaviour?
If so, I can finish the PR, improve the details and submit it 

PR: https://github.com/RocketChat/Rocket.Chat/pull/21065
Demo: https://streamable.com/p0cxfj

**Comment by Julianoe** [comment]
For me that's one good UX take on that issue 👍
